IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m

```diff
@@ -880,32 +880,67 @@ - (BOOL)saveAppMetadataWithConfiguration:(MSIDConfiguration *)configuration
                                                           error:(NSError *__autoreleasing *)error
 {
-    NSString *environment = [[configuration.authority cacheUrlWithContext:context] msidHostWithPortIfNecessary];
-    MSIDAppMetadataCacheKey *metadataKey = [[MSIDAppMetadataCacheKey alloc] initWithClientId:configuration.clientId
-                                                                                 environment:environment
-                                                                                    familyId:nil
-                                                                                 generalType:MSIDAppMetadataType];
+    MSIDAppMetadataCacheQuery *metadataQuery = [[MSIDAppMetadataCacheQuery alloc] init];
+    metadataQuery.clientId = configuration.clientId;
+    metadataQuery.generalType = MSIDAppMetadataType;
+    metadataQuery.environmentAliases = [configuration.authority defaultCacheEnvironmentAliases];
+    return [_accountCredentialCache getAppMetadataWithQuery:metadataQuery context:context error:error];
+}
+
+- (BOOL)removeAppMetadata:(MSIDAppMetadataCacheItem *)appMetadata
+                  context:(id<MSIDRequestContext>)context
+                    error:(NSError **)error
+{
+    if (!appMetadata)
+    {
+        [self fillInternalErrorWithMessage:@"App metadata not provided, cannot remove" context:context error:error];
+        return NO;
+    }
     
-    return [_accountCredentialCache getAppMetadata:metadataKey context:context error:error];
+    MSIDTelemetryCacheEvent *event = [MSIDTelemetry startCacheEventWithName:MSID_TELEMETRY_EVENT_APP_METADATA_DELETE
+                                                                    context:context];
+    BOOL result = [_accountCredentialCache removeAppMetadata:appMetadata context:context error:error];
+    [MSIDTelemetry stopCacheEvent:event withItem:nil success:result context:context];
+    return result;
+}
+
+- (BOOL)updateAppMetadata:(MSIDAppMetadataCacheItem *)appMetadata
+                  context:(id<MSIDRequestContext>)context
+                    error:(NSError **)error
+{
+    MSIDTelemetryCacheEvent *event = [MSIDTelemetry startCacheEventWithName:MSID_TELEMETRY_EVENT_APP_METADATA_WRITE
+                                                                    context:context];
+    BOOL result = [_accountCredentialCache saveAppMetadata:appMetadata context:context error:error];
+    [MSIDTelemetry stopCacheEvent:event withItem:nil success:result context:context];
+    return result;
+}
+
+- (NSArray<MSIDAppMetadataCacheItem *> *)getAppMetadataEntries:(MSIDConfiguration *)configuration
+                                                       context:(id<MSIDRequestContext>)context
+                                                         error:(NSError *__autoreleasing *)error
+{
+    MSIDAppMetadataCacheQuery *metadataQuery = [[MSIDAppMetadataCacheQuery alloc] init];
```

Comment:
Shouldn't we have a metadata reading telemetry event?

Reply:
I think we can take that as a separate PR. The current MSIDTelemetry stopCacheEvent takes only MSIDBaseToken as parameter. We will need to make that api accept an id and make other changes as well.

Reply:
Sounds good to me. Please create a github issue for it and we can take it later.